cpp/src/arrow/CMakeLists.txt

```diff
@@ -69,6 +66,14 @@ set(ARROW_SRCS
   type.cc
   visitor.cc
 
+  columnar/builder_adaptive.cc
```

Comment:
Hmm... Honestly "columnar" doesn't mean anything to me. Why would it have builder classes but not Table (i.e. an actual columnar container)?

Reply:
Can you suggest another name? "core" perhaps?

Reply:
What do you intend to put in it?

Reply:
I would like to have a place to organize the columnar data structure implementation details in a way that is logical for developers (of the Arrow codebase, not users of its public API) to navigate, and that promotes developer productivity (through easier navigation, smaller headers and compilation units, etc.). This should be unrelated to the development of compact public API (we should put a bit more effort into organizing that also -- it has grown a tad crufty, or at least it not well-documented)

The next thing to do would be to organize array.h/array.cc into similarly-named files. Parts of other files could be moved there, too. I couldn't bear to keep adding more classes to builder.h hence the decision to organize a bit here

Reply:
Call it `array` simply? We'd put all the `array-*` and `builder-*` stuff inside.

Reply:
done